### PR TITLE
Merge LitterType into the shop item descriptor

### PR DESCRIPTION
**Pull request by duncanspumpkin**
Continuing on from #13311

This merges a bunch of shop item information into the ShopItemDescriptor and tides up the peep code to use the 64bit versions of the flags to reduce code duplication.

I've noticed an original bug in the consumption code where if you have food and drink the game will assume that you are consuming a drink and only update the peeps thirst.

### CODE REVIEW

File: src/openrct2/peep/Guest.cpp [hunk 1]
```diff
@@ -763,36 +589,21 @@ void Guest::loc_68FA89()
 
         if (TimeToConsume == 0)
         {
-            int32_t chosen_food = bitscanforward(HasFoodStandardFlag());
+            int32_t chosen_food = bitscanforward(GetFoodFlags());
```
Comment (duncanspumpkin) [inline src/openrct2/peep/Guest.cpp:592]: Why is this not consuming Drink?

File: src/openrct2/peep/Guest.cpp [hunk 2]
```diff
@@ -1028,12 +839,12 @@ void Guest::Tick128UpdateGuest(int32_t index)
...
                     }
 
-                    if (Hunger <= 10 && !HasFood())
+                    if (Hunger <= 10 && !GetFoodFlags())
                     {
                         possible_thoughts[num_thoughts++] = PEEP_THOUGHT_TYPE_HUNGRY;
                     }
 
-                    if (Thirst <= 25 && !HasFood())
+                    if (Thirst <= 25 && !GetFoodFlags())
```
Comment (duncanspumpkin) [inline src/openrct2/peep/Guest.cpp:847]: Drink?

File: src/openrct2/peep/Guest.cpp [hunk 3]
```diff
@@ -1202,7 +1013,7 @@ void Guest::Tick128UpdateGuest(int32_t index)
 void Guest::TryGetUpFromSitting()
 {
     // Eats all food first
-    if (HasFood())
+    if (GetFoodFlags())
```
Comment (duncanspumpkin) [inline src/openrct2/peep/Guest.cpp:1016]: What about drinks?

File: src/openrct2/peep/Guest.cpp [hunk 4]
```diff
@@ -1280,7 +1091,7 @@ void Guest::UpdateSitting()
             return;
         }
 
-        if (HasFood())
+        if (GetFoodFlags())
```
Comment (duncanspumpkin) [inline src/openrct2/peep/Guest.cpp:1094]: Again drinks?

File: src/openrct2/peep/Guest.cpp [hunk 5]
```diff
@@ -1485,15 +1268,10 @@ bool Guest::DecideAndBuyItem(Ride* ride, ShopItem shopItem, money32 price)
...
     if (GetShopItemDescriptor(shopItem).IsFoodOrDrink())
     {
-        int32_t food = -1;
-        if ((food = HasFoodStandardFlag()) != 0)
-        {
-            InsertNewThought(PEEP_THOUGHT_TYPE_HAVENT_FINISHED, bitscanforward(food));
-            return false;
-        }
-        else if ((food = HasFoodExtraFlag()) != 0)
+        int32_t food = bitscanforward(GetFoodFlags());
```
Comment (duncanspumpkin) [inline src/openrct2/peep/Guest.cpp:1271]: Again drink

File: src/openrct2/peep/Guest.cpp [hunk 6]
```diff
@@ -1831,7 +1614,7 @@ void Guest::PickRideToGoOn()
         return;
     if (PeepFlags & PEEP_FLAGS_LEAVING_PARK)
         return;
-    if (HasFood())
+    if (GetFoodFlags())
```
Comment (duncanspumpkin) [inline src/openrct2/peep/Guest.cpp:1617]: or drink?

File: src/openrct2/peep/Guest.cpp [hunk 7]
```diff
@@ -3243,7 +3026,7 @@ static void peep_head_for_nearest_ride_type(Guest* peep, int32_t rideType)
 
 static void peep_head_for_nearest_ride_with_flags(Guest* peep, int32_t rideTypeFlags)
 {
-    if ((rideTypeFlags & RIDE_TYPE_FLAG_IS_TOILET) && peep->HasFood())
+    if ((rideTypeFlags & RIDE_TYPE_FLAG_IS_TOILET) && peep->GetFoodFlags())
```
Comment (duncanspumpkin) [inline src/openrct2/peep/Guest.cpp:3029]: or drink

**Comment by duncanspumpkin** [comment]
Consider renaming GetFoodFlags to GetFoodOrDrinkFlags to match the shop item descriptor. HasDrink can be removed if logic is reworked to correct above mentioned bug.

File: src/openrct2/ride/ShopItem.h [hunk 8]
```diff
@@ -98,27 +93,39 @@ struct ShopItemDescriptor
     uint32_t Image;
     ShopItemStrings Naming;
     uint16_t Flags;
+    uint8_t LitterType;
+    uint8_t ConsumptionTime;
+    ShopItem DiscardContianer;
```
Comment (tupaschoal) [inline src/openrct2/ride/ShopItem.h:98]:
```suggestion
    ShopItem DiscardContainer;
```

File: src/openrct2/ride/ShopItem.h [hunk 9]
```diff
@@ -98,27 +93,39 @@ struct ShopItemDescriptor
...
+    PeepThoughtType TooMuchThought;
+    PeepThoughtType GoodValueThought;
 
-    bool HasFlag(const uint16_t flag) const;
+    constexpr bool HasFlag(const uint16_t flag) const
```
Comment (tupaschoal) [inline src/openrct2/ride/ShopItem.h:102]: Why moving this to the header?

Comment (duncanspumpkin) [reply src/openrct2/ride/ShopItem.h:102]: constexpr need them in the header. Its the reason why i've got a bit of a convoluted thing to get all the flags out of the array.

File: src/openrct2/ride/ShopItem.h (hunk 8 above)
Comment (duncanspumpkin) [reply src/openrct2/ride/ShopItem.h:98]: Knew it didn't look correct.

**Review by tupaschoal — APPROVED** [review]

File: src/openrct2-ui/windows/Guest.cpp [hunk 10]
```diff
@@ -1982,7 +1982,7 @@ void window_guest_inventory_paint(rct_window* w, rct_drawpixelinfo* dpi)
     gfx_draw_string_left(dpi, STR_CARRYING, nullptr, COLOUR_BLACK, screenCoords);
     screenCoords.y += 10;
 
-    for (ShopItem item = ShopItem::Balloon; item < ShopItem::Count; item++)
+    for (ShopItem item = ShopItem::Balloon; item < ShopItem::Count; item = ShopItem(EnumValue(item) + 1))
```
Comment (Gymnasiast) [inline src/openrct2-ui/windows/Guest.cpp:1985]: Is it not possible to implement `++` for ShopItem?

**Comment by Gymnasiast** [comment]
I´m not sure how it ended up with that odd capitalisation, but could you change `ShopItem::SuJeongGwa` to `ShopItem::Sujeonggwa`?

File: src/openrct2-ui/windows/Guest.cpp (hunk 10 above)
Comment (duncanspumpkin) [reply src/openrct2-ui/windows/Guest.cpp:1985]: I tend to not like having ++ for enum classes as you can get values that are not valid. For example this loop will include a couple invalid items as not all 64 bits are in use.

Comment (Gymnasiast) [reply src/openrct2-ui/windows/Guest.cpp:1985]: Well, exactly - this has the same downside, except it looks very ugly on top.

Can we define a ++ that skips these invalid values?

**Review by Gymnasiast — APPROVED** [review]

**Pull request merged**